### PR TITLE
AdaFactor optimizer: memory-efficient adaptive optimizer

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AdaFactor is a memory-efficient adaptive optimizer that uses factored second-moment estimates, requiring ~2x less memory than Adam/AdamW. For large-capacity DrivAerML models (4L/512d/8H), AdaFactor's factored approximation may act as implicit regularization that improves generalization — particularly because the 3D car geometry dataset has limited training samples (~500 cars). Fixed-LR mode with a constant learning rate decouples AdaFactor from its aggressive internal LR schedule, making it a fair comparison to AdamW.

Note: This was originally assigned in Wave 16 (would-have-been PR #3036) but the branch never launched. Clean re-assignment from radford baseline.

**Cross-dataset coverage:** TandemFoil (TF), TandemFoil Paper (TFP), AirfRANS (AF), DrivAerML (DM).

## Current Baselines

| Dataset | Metric | Value | PR |
|---------|--------|-------|----|
| TF | `val_primary/surface_pressure_mae` | 22.537 | #2924 |
| TFP | `val_primary/field_mse` | 0.00434 | #2979 |
| AF | `val_primary/surface_mse` | 0.000340 | #2898 |
| DM | `val_primary/surface_rel_l2_pct` | 3.997% | #2898 |

**External DrivAerML target:** <3.71% (AB-UPT benchmark)

## Implementation

Add `--optimizer adafactor` option to `train.py`. Use `transformers.Adafactor` or `torch_optimizer.Adafactor`:

```python
from transformers import Adafactor

optimizer = Adafactor(
    model.parameters(),
    lr=args.lr,           # fixed LR mode: pass explicit lr
    scale_parameter=False,  # disable internal LR scaling
    relative_step=False,    # disable relative step sizing
    warmup_init=False,      # no warmup (we handle externally)
    weight_decay=args.weight_decay,
    clip_threshold=1.0,     # built-in gradient clipping
)
```

**Key difference from AdamW:** AdaFactor stores a factored approximation of the second moment: for a 2D weight matrix `W [m x n]`, it stores `R [m]` and `C [n]` instead of a full `[m x n]` matrix — O(m+n) vs O(m*n) memory. For large layers (512d hidden), this is a significant saving.

**Use fixed LR mode:** Pass `lr=args.lr, scale_parameter=False, relative_step=False`. This is essential — AdaFactor's default relative step mode uses an aggressive internal scheduler that will fight our cosine annealing.

## Experiment Instructions

Run the following experiments (GPU allocation: DM=4 GPUs, TF=2 GPUs, TFP=1 GPU, AF=1 GPU):

**TandemFoil (2 GPUs) — use Lion LR equivalent:**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer adafactor --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb-group adafactor-optimizer
```

**TandemFoil Paper (1 GPU):**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adafactor --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --wandb-group adafactor-optimizer
```

**AirfRANS (1 GPU):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adafactor --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group adafactor-optimizer
```

**DrivAerML (4 GPUs):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adafactor --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group adafactor-optimizer
```

## Expected Outcome

- DM: improvement from 3.997% (factored second moments act as regularization; implicit per-parameter LR smoothing)
- TFP: improvement from 0.00434 (memory saving enables larger effective batch or minor generalization gain)
- TF and AF: comparable or slight regression vs Lion/AdamW (Lion is already tuned for TF; AdamW is tuned for AF)

## Suggested Follow-ups

- If AdaFactor underperforms AdamW at lr=5e-4, try lr=1e-3 (AdaFactor can tolerate higher LRs due to internal normalization)
- If factored approximation hurts on small models (TF/TFP with 192d), try AdaFactor only on DM+AF
- Consider AdaFactor with `scale_parameter=True` (internal LR mode) as an ablation if fixed-LR mode fails

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
